pkg/common/cns-lib/volume/manager.go

```diff
@@ -2323,9 +2332,44 @@ func (m *defaultManager) createSnapshotWithImprovedIdempotencyCheck(ctx context.
 		// Local instance of CreateSnapshot details that needs to be persisted.
 		volumeOperationDetails *cnsvolumeoperationrequest.VolumeOperationRequestDetails
 		// error
-		err error
+		err                        error
+		quotaInfo                  *cnsvolumeoperationrequest.QuotaDetails
+		isStorageQuotaM2FSSEnabled bool
 	)
+	if extraParams != nil {
+		createSnapParams, ok := extraParams.(*CreateSnapshotExtraParams)
+		if !ok {
+			return nil, logger.LogNewErrorf(log, "unrecognised type for CreateSnapshot params: %+v", extraParams)
+		}
+		log.Debugf("Received CreateVolume extraParams: %+v", *createSnapParams)
 
+		isStorageQuotaM2FSSEnabled = createSnapParams.IsStorageQuotaM2FSSEnabled
+		if isStorageQuotaM2FSSEnabled && m.clusterFlavor == cnstypes.CnsClusterFlavorWorkload {
+			var storagePolicyID string
+			queryFilter := cnstypes.CnsQueryFilter{
+				VolumeIds: []cnstypes.CnsVolumeId{{Id: volumeID}},
+			}
+			queryResult, err := m.QueryVolume(ctx, queryFilter)
+			if err != nil {
+				return nil, logger.LogNewErrorf(log,
+					"queryVolume failed for queryFilter: %+v. Err=%+v", queryFilter, err.Error())
+			}
+			if len(queryResult.Volumes) == 0 {
+				return nil, logger.LogNewErrorf(log,
+					"volumeID %q not found in QueryVolume", volumeID)
+			}
+			storagePolicyID = queryResult.Volumes[0].StoragePolicyId
+			volSize := queryResult.Volumes[0].BackingObjectDetails.GetCnsBackingObjectDetails().CapacityInMb
```

Comment:
We can optimize this imo, perhaps cnsvolumeinfo itself can hold the size of the source PVC, this can be populated during createvolume call and updated during extendvolume calls. However, this is a DB call, should complete fast imo. In anycase just add a task into the execution page, we'll pick it up if we hit perf bottlenecks.

Reply:
cnsvolumeinfo holds the size of source PVC. no need for fetching volume again.